### DISCUSSION TITLE
TakeSnapShot

### DISCUSSION THREAD

**Discussion by firemen01** [FactoryXCode/MfPack#49] [Q&A]
create timer interval 50, 

at the start of video playback, memory begins to grow and after a while the rendering stops.
what am I doing wrong?
it feels like the memory is not being cleared and there is a leak somewhere

```
procedure Tfrm_MfPlayer.Timer1Timer(Sender: TObject);
var
  bm: Tbitmap;
begin
    if not MfPlayerPlayVideo then exit;

    bm:= TBitmap.Create;

    MfPlayer.TakeSnapShot(bm);

    myTestPanel.Canvas.lock;
    myTestPanel.Canvas.Draw(0,0, bm);
    myTestPanel.Canvas.Unlock;

    FreeAndNil(bm);

end;
```

**Comment by firemen01** [comment]
the result is the same if you make the global bitmap in create \free along with the main form

**Reply by FactoryXCode** [reply]
I suggest you have a look at MfPlayer.TakeSnapShot(bm);

Although you have  a few things to consider:

1 A TTimer is not suitable for what you want to do. The timer messages are to slow to process through the system, because timer messages have low priority in the Windows message queue. In Short: While your timer is launching messages, they are most of the time not in sync with all messages processed by the engine, that works A-synchronous, they come to late or to early, depending on the consumed time a method use. In that case it's possible that sources are not released in time, what explains the grow of memory.  A way to comfort is to stop the timer in OnTimer event, process the code and start TTimer again in the OnTimer so all processes are done.
2 The player works a-synchronous, This means you have to wait until all processes, like freeing the bitmaps are done. For this you should implement a method that waits, until the given processes are done, like WaitForSingleObject [See here](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject)
3 Setting a timer to an arbitrary 50 is not in synch with the framerate of the reader. If this is for example is 30 frames/second, your timer needs to be synchronized with 1/30 of a second.. And that is not possible, because the timer does not know the time elapse the reader is processing. A possible solution is to send a message to your mainform or caller, when each sample is processed, and then take a snapshot (not easy too, but worth trying).
